### PR TITLE
JPile retries. No dependencies added.

### PR DESCRIPTION
**Pull request by io-german**
This is an alternative to https://github.com/opower/jpile/pull/45

```JdbcUtil``` class was refactored to ```ConnectionBasedStatementExecutor``` class implementing ```StatementExecutor``` interface. Now any client can implement required functionality on his side by implementing ```StatementExecutor``` interface.  

```ConnectionBasedStatementExecutor``` and ```DataSourceBasedStatementExecutor``` can be used as-is or as a base class for client implementations.

Old API was deprecated but still available so JPile users have time to switch to new implementations if they require it.

### CODE REVIEW

File: src/main/java/com/opower/persistence/jpile/jdbc/DataSourceBasedStatementExecutor.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+     */
+    @Override
+    public void shutdown() {
+        if (this.connectionBasedStatementExecutor != null) {
+            this.connectionBasedStatementExecutor.closeConnection();
+            this.connectionBasedStatementExecutor.shutdown();
+        }
+    }
+
+    private synchronized void enableConnection() throws SQLException {
```
Comment (ph1lm) [inline src/main/java/com/opower/persistence/jpile/jdbc/DataSourceBasedStatementExecutor.java]: I would renamed it to **re**enableConnection()

File: src/main/java/com/opower/persistence/jpile/jdbc/DataSourceBasedStatementExecutor.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+                this.connectionBasedStatementExecutor, "underlying connectionBasedStatementExecutor is null");
+
+        return this.connectionBasedStatementExecutor.execute(statementCallback);
+    }
+
+    /**
+     * @inheritDoc
+     */
+    @Override
+    public void shutdown() {
```
Comment (ph1lm) [inline src/main/java/com/opower/persistence/jpile/jdbc/DataSourceBasedStatementExecutor.java]: ``synchronized``

File: src/main/java/com/opower/persistence/jpile/jdbc/DataSourceBasedStatementExecutor.java (hunk 1 above)
Comment (io-german) [inline src/main/java/com/opower/persistence/jpile/jdbc/DataSourceBasedStatementExecutor.java]: or ```initNewConnection()```

Comment (io-german) [inline src/main/java/com/opower/persistence/jpile/jdbc/DataSourceBasedStatementExecutor.java]: Done

File: src/main/java/com/opower/persistence/jpile/jdbc/DataSourceBasedStatementExecutor.java (hunk 2 above)
Comment (io-german) [inline src/main/java/com/opower/persistence/jpile/jdbc/DataSourceBasedStatementExecutor.java]: Done

File: src/main/java/com/opower/persistence/jpile/jdbc/ConnectionBasedStatementExecutor.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    }
+
+    /**
+     * Force close the connection. This method is redundant for {@link StatementExecutor} interface
+     * but can be used for other {@link StatementExecutor} that uses this exact implementation
+     * (e.g. {@link DataSourceBasedStatementExecutor}).
+     */
+    public void closeConnection() {
+        try {
+            if (this.connection != null) {
```
Comment (ph1lm) [inline src/main/java/com/opower/persistence/jpile/jdbc/ConnectionBasedStatementExecutor.java]: How it could be null?

Comment (io-german) [inline src/main/java/com/opower/persistence/jpile/jdbc/ConnectionBasedStatementExecutor.java]: It can't. Fixed.

**Comment by amir20** [comment]
This is awesome. Thanks for doing it.

File: src/main/java/com/opower/persistence/jpile/jdbc/DataSourceBasedStatementExecutor.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    private ConnectionBasedStatementExecutor connectionBasedStatementExecutor;
+    private final DataSource dataSource;
+
+    public DataSourceBasedStatementExecutor(DataSource dataSource) throws SQLException {
+        Preconditions.checkNotNull(dataSource, "dataSource must not be null");
+
+        this.dataSource = dataSource;
+    }
+
+
```
Comment (ph1lm) [inline src/main/java/com/opower/persistence/jpile/jdbc/DataSourceBasedStatementExecutor.java]: redundant new lines

File: src/main/java/com/opower/persistence/jpile/jdbc/DataSourceBasedStatementExecutor.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+    /**
+     * @inheritDoc
+     */
+    @Override
+    public synchronized <T> T execute(StatementCallback<T> statementCallback) {
+        if (this.connectionBasedStatementExecutor == null) {
```
Comment (ph1lm) [inline src/main/java/com/opower/persistence/jpile/jdbc/DataSourceBasedStatementExecutor.java]: I would make separate synchronized method for this.

File: src/main/java/com/opower/persistence/jpile/jdbc/DataSourceBasedStatementExecutor.java (hunk 4 above)
Comment (io-german) [inline src/main/java/com/opower/persistence/jpile/jdbc/DataSourceBasedStatementExecutor.java]: Done

File: src/main/java/com/opower/persistence/jpile/jdbc/DataSourceBasedStatementExecutor.java (hunk 5 above)
Comment (io-german) [inline src/main/java/com/opower/persistence/jpile/jdbc/DataSourceBasedStatementExecutor.java]: Done

**Comment by ph1lm** [comment]
lgtm

**Comment by io-german** [comment]
@amir20 @corcorb Could you please merge this PR if you don't have any new comments. 
Thanks.

**Comment by io-german** [comment]
@amir20 @corcorb ping

**Comment by amir20** [comment]
This one looks good. I'll merge.

**Pull request merged**